rfcs/profiler.md

```diff
@@ -0,0 +1,144 @@
+- Our solution should be able to aggregate metrics from multiple components into a roll-up view for internal use (may be useful for the third party as well)
+- ~~[P2] Our solution should be able to be released to clients to help them identify and improve poor performing queries~~
+
+## Future Work
+
+This work is open ended as there is a chance to add more and more custom profilers to the system. Also right now, we plan to use flamegraph to connect custom profilers and the generic main profiler. This can also be extended by using different intermediate formats or objects.
+
+## Open Sourcing Strategy
+
+This work can be easily open sourced as the profiling is a very general tool (both generic and custom ones) and can benefit any computationally heavy service or infrastructure project.
+
+# Design
+
+Given the complexity of the system and layers of software involved, we are taking a hierarchical approach (multilayer) to implement an efficient and useful profiler:
+
+## Generic Profiler
+
+To address the mentioned basic requirements, at the top level, we use a generic and simple profiler which only tracks end-to-end execution time and memory metrics (peak memory, page faults, etc.) of the executed queries. This profiler uses a commonly available tool such as **_time_** to keep track of this high level information across days and releases in a small database (the early version can be using just the filesystem (key will be the filename and value will be the file content) and for then next phase, we can use [SqliteDict](https://pythonhosted.org/sqlite_object/sqlite_dict.html) K/V database, TileDB or DynamoDB as other alternatives) across different platforms. As the profiler runs the application process, a new record will be generated and stored. The record associated with each run has the following schema:
+
+- **Key:**
+  - _Date_
+  - _Time_
+  - _Process Information (application and its arguments)_
+- **Value:**
+  - _Host information_
+  - _User Defined MD_
+    - _Libraries version_
+    - _Git commit SHA1_
+    - _Description_
+    - _TileDB stats_
+  - _Tag_
+  - _Time (real time, sys time, user time)_
+  - _Memory (peak, max resident, page faults, …)_
+  - _Custom Profiler 1 output_
+  - _Custom Profiler 2 output_
+
+As can be seen, the key is a combination of **_Date_**, **_Time_** and **_Process Info_**. The idea is to make it easy to query across time for a given process. The value consists of executed process detail info, library versions (releases and PR perhaps too). Additionally, the user can add platform information (like Single Cell library versions) and git **_SHA1_** and a _description_ for the benchmark. **_Tag_** is a value linking multiple benchmarks (a suite of runs). Finally, the profile information also includes time and memory metrics. Additionally, it includes the output info of custom profilers. Custom profilers will be discussed in the next section.
```

Comment:
```suggestion
As can be seen, the key is a combination of **_Date_**, **_Time_** and **_Process Info_**. The idea is to make it easy to query across time for a given process. The value consists of executed process detail info, library versions (releases and PR perhaps too). Additionally, the user can add platform information (like SOMA library versions) and git **_SHA1_** and a _description_ for the benchmark. **_Tag_** is a value linking multiple benchmarks (a suite of runs). Finally, the profile information also includes time and memory metrics. Additionally, it includes the output info of custom profilers. Custom profilers will be discussed in the next section.
```

Reply:
The spec branding is SOMA. The impl branding is TileDB-SOMA. There is no "Single Cell" branding in our terminology.